### PR TITLE
Extract StoresResources from Harvester

### PR DESCRIPTION
**Pull request by PunkPun**
This is the first PR of the resource refactor. 

- Renamed StoresResources -> StoresPlayerResources
- Extracted StoresResources from Harvester (StoresResources diff is messed up)
- Renamed WithHarvesterPipsDecoration -> WithStoresResourcesPipsDecoration

This was an easy first step that could be done before docking refactor or the massive systematic changes to player resources.

With this PR harvesters will be able to have multiple storage containers, although the pips will be shown for only the one (consistent with most of our UI traits). We may want to create a way of tying those traits together somehow in the future.

We may also want to consider adding per-resource conditions so enable features such as TS harvester exploding when containing blue tib. I can add this feature in this PR if requested.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Render/WithStoresResourcesPipsDecoration.cs [hunk 1]
```diff
@@ -42,26 +42,27 @@ public class WithHarvesterPipsDecorationInfo : WithDecorationBaseInfo, Requires<
...
+		readonly IStoresResources storesResources;
 		readonly Animation pips;
 
-		public WithHarvesterPipsDecoration(Actor self, WithHarvesterPipsDecorationInfo info)
+		public WithStoresResourcesPipsDecoration(Actor self, WithStoresResourcesPipsDecorationInfo info)
 			: base(self, info)
 		{
-			harvester = self.Trait<Harvester>();
+			// Will never be null as we require this trait.
+			storesResources = self.TraitOrDefault<IStoresResources>();
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithStoresResourcesPipsDecoration.cs]:
```suggestion
			storesResources = self.Trait<IStoresResources>();
```
?

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 2]
```diff
@@ -170,8 +170,28 @@ public interface IVoiced
...
+	}
+
+	public interface IStoresResources
+	{
+		bool HasType(string resourceType);
+
+		/// <summary>The amount of resources that can stored.</summary>
+		int Capacity { get; }
+
+		/// <summary>Performance cheep method of getting the total sum of contents.</summary>
```
Comment (Mailaender) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]:
```suggestion
		/// <summary>Performance wise cheap method of getting the total sum of contents.</summary>
```

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 3]
```diff
@@ -170,8 +170,28 @@ public interface IVoiced
...
+		int TotalContents { get; }
+		IReadOnlyDictionary<string, int> Contents { get; }
+
+		/// <summary>If accepted returns 0. Otherwise returns however much value was unnacepted.</summary>
```
Comment (Mailaender) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]:
```suggestion
		/// <summary>If accepted returns 0. Otherwise returns much was not accepted.</summary>
```

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 4]
```diff
@@ -170,8 +170,28 @@ public interface IVoiced
...
+		int AddResource(string resourceType, int value);
+
+		/// <summary>If consumed returns 0. Otherwise returns however much value was unconsumed.</summary>
```
Comment (Mailaender) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]:
```suggestion
		/// <summary>If consumed returns 0. Otherwise returns how much was not consumed.</summary>
```

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 5]
```diff
@@ -170,8 +170,28 @@ public interface IVoiced
...
+	public interface IStoresResourcesInfo : ITraitInfoInterface
+	{
+		string[] ResourceTypes { get; }
+	}
+
+	public interface IStoresResources
+	{
+		bool HasType(string resourceType);
+
+		/// <summary>The amount of resources that can stored.</summary>
```
Comment (Mailaender) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]:
```suggestion
		/// <summary>The amount of resources that can be stored.</summary>
```

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 6]
```diff
@@ -170,8 +170,31 @@ public interface IVoiced
...
+	public interface IStoresResources
+	{
+		bool HasType(string resourceType);
+
+		/// <summary>The amount of resources that can be stored.</summary>
+		int Capacity { get; }
+
+		/// <summary>Stored resources.</summary>
+		/// <remarks>Dictionary key refers to resourceType, value refers to resource amount.</remarks>
+		IReadOnlyDictionary<string, int> Contents { get; }
```
Comment (anvilvapre) [inline OpenRA.Game/Traits/TraitsInterfaces.cs:190]: I understand you want to keep resource type a string. 

On the other hand I assume this is a hot path. Seems somewhat wastefull to do constant string hash calculations and hash map lookups to get and set an integer value.

HarvesterInfo.Resources is readonly. You could assign integer indexes. So that you can get rid of of both the string and map and use an array of length Resources.

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 7]
```diff
@@ -102,17 +98,32 @@ public class HarvesterInfo : ConditionalTraitInfo, Requires<MobileInfo>
...
+			{
+				foreach (var type in sr.ResourceTypes)
+				{
+					if (Resources.Contains(type) && resourceTypes.Contains(type))
+						resourceTypes.Remove(type);
+				}
+			}
+
+			if (resourceTypes.Count != 0)
+				throw new YamlException($"Harvester can't store resources: {string.Join(',', resourceTypes)} ");
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: A harvester should be able to handle all resource types? It couldn't be some other actor type that handles other resource types?

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 8]
```diff
@@ -216,9 +215,9 @@ public Actor ClosestProc(Actor self, Actor ignore)
 			return null;
 		}
 
-		public bool IsFull => contents.Values.Sum() == Info.Capacity;
-		public bool IsEmpty => contents.Values.Sum() == 0;
-		public int Fullness => contents.Values.Sum() * 100 / Info.Capacity;
+		public bool IsFull => storesResources.All(sr => sr.ContentsSum == sr.Capacity);
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: `>=` to be save?

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 9]
```diff
@@ -216,9 +215,9 @@ public Actor ClosestProc(Actor self, Actor ignore)
...
+		public bool IsEmpty => storesResources.All(sr => sr.ContentsSum == 0);
+		public int Fullness => storesResources.Sum(sr => sr.ContentsSum * 100 / sr.Capacity) / storesResources.Length;
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/Harvester.cs:126]: `storedResources.Length` might be 0. Shouldn't be, but could be? Check in constructor?

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 10]
```diff
@@ -250,28 +248,27 @@ public virtual bool TickUnload(Actor self, Actor proc)
...
+					var accepted = acceptResources.AcceptResources(c.Key, count);
 					if (accepted == 0)
 						continue;
 
-					contents[resourceType] -= accepted;
-					if (contents[resourceType] <= 0)
-						contents.Remove(resourceType);
-
+					sr.RemoveResource(c.Key, accepted);
 					currentUnloadTicks = Info.BaleUnloadDelay;
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/Harvester.cs:181]: No need to adjust this to `accepted` value? I.e. unload partial bales faster?

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 11]
```diff
@@ -250,28 +248,27 @@ public virtual bool TickUnload(Actor self, Actor proc)
...
 					UpdateCondition(self);
 					return false;
 				}
 			}
 
-			return contents.Count == 0;
+			return IsEmpty;
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/Harvester.cs:187]: Seems strange/buggy, already before? In the above loop you try to find resources to unload. If not you exit or don't enter the loop, and then here again you check the total resources to again determine if empty?

Shouldn't this just always be `true`?

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 12]
```diff
@@ -368,14 +365,14 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 
 		int ISpeedModifier.GetSpeedModifier()
 		{
-			return 100 - (100 - Info.FullyLoadedSpeed) * contents.Values.Sum() / Info.Capacity;
+			return 100 - (100 - Info.FullyLoadedSpeed) * Fullness / 100;
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/Harvester.cs:269]: integer division. can't tell if this will always go right?

File: OpenRA.Mods.Common/Traits/Render/WithStoresResourcesPipsDecoration.cs [hunk 13]
```diff
@@ -42,26 +42,26 @@ public class WithHarvesterPipsDecorationInfo : WithDecorationBaseInfo, Requires<
...
-		readonly Harvester harvester;
+		readonly IStoresResources storesResources;
 		readonly Animation pips;
 
-		public WithHarvesterPipsDecoration(Actor self, WithHarvesterPipsDecorationInfo info)
+		public WithStoresResourcesPipsDecoration(Actor self, WithStoresResourcesPipsDecorationInfo info)
 			: base(self, info)
 		{
-			harvester = self.Trait<Harvester>();
+			storesResources = self.Trait<IStoresResources>();
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/Render/WithStoresResourcesPipsDecoration.cs]: Plural is a bit confusing. `storeResourcesTrait` seems clearer. Harvester used to be clearer.

File: OpenRA.Mods.Common/Traits/StoresPlayerResources.cs [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+ */
+#endregion
+
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Adds capacity to a player's harvested resource limit.")]
+	public class StoresPlayerResourcesInfo : TraitInfo
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/StoresPlayerResources.cs:18]: Fine. Just makes me wonder. Is it a resource of the actor, player or map. 
Is it a single resource type with a (monetary) value. Or is each i.e. gem an instance of resource that you might like to store by itself.

File: OpenRA.Mods.Common/Traits/StoresPlayerResources.cs [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+		public override object Create(ActorInitializer init) { return new StoresPlayerResources(init.Self, this); }
+	}
+
+	public class StoresPlayerResources : INotifyOwnerChanged, INotifyCapture, ISync, INotifyKilled, INotifyAddedToWorld, INotifyRemovedFromWorld
+	{
+		readonly StoresPlayerResourcesInfo info;
+		PlayerResources player;
+
+		[Sync]
+		public int Stored => player.ResourceCapacity == 0 ? 0 : (int)((long)info.Capacity * player.Resources / player.ResourceCapacity);
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/StoresPlayerResources.cs:31]: Regarding the sync. Why just just two integer fields, one `PlayerCapacity` and `Player.Resources`. Skip the calculations. 

If player already syncs PlayerCapacity, then you can just sync player.Resources?

Why even sync this, if those player properties also are synched?

File: OpenRA.Mods.Common/Traits/StoresPlayerResources.cs [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+		void INotifyKilled.Killed(Actor self, AttackInfo e)
+		{
+			// Lose the stored resources
+			player.TakeResources(Stored);
+		}
+
+		void INotifyAddedToWorld.AddedToWorld(Actor self)
+		{
+			player.AddStorage(info.Capacity);
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/StoresPlayerResources.cs]: Storage? `SetResourceCapacity`?

File: OpenRA.Mods.Common/Traits/StoresResources.cs [hunk 17]
```diff
@@ -9,62 +9,100 @@
...
-using OpenRA.Primitives;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Adds capacity to a player's harvested resource limit.")]
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/StoresResources.cs]: harvested/stored/storable

File: OpenRA.Mods.Common/Traits/StoresResources.cs [hunk 18]
```diff
@@ -9,62 +9,100 @@
...
+			{
+				var value = 0;
+				foreach (var c in contents)
+					value += c.Value << c.Key.Length;
+
+				return value;
+			}
+		}
+
+		public int ContentsSum { get; private set; } = 0;
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/StoresResources.cs:53]: Expensive sync.

File: OpenRA.Mods.Common/Traits/StoresResources.cs [hunk 19]
```diff
@@ -9,62 +9,100 @@
...
+			return info.Resources.Contains(resourceType);
 		}
 
-		void INotifyKilled.Killed(Actor self, AttackInfo e)
+		int IStoresResources.AddResource(string resourceType, int value)
 		{
-			// Lose the stored resources
-			player.TakeResources(Stored);
-		}
+			if (!HasType(resourceType))
```
Comment (anvilvapre) [inline OpenRA.Mods.Common/Traits/StoresResources.cs:74]: `Game.World.GetIdForResourceType(resourceType)`

**Review by anvilvapre — DISMISSED** [review]
untested. 

`StoresPlayerResourceAmount`, seems better than `StoresPlayerResources`.

Or `PlayerResourceConsumer`.

File: OpenRA.Mods.Common/Traits/StoresPlayerResources.cs (hunk 16 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/StoresPlayerResources.cs]: Makes sense

File: OpenRA.Mods.Common/Traits/Harvester.cs (hunk 9 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Harvester.cs:126]: Added a rulesetloaded check

File: OpenRA.Mods.Common/Traits/Harvester.cs (hunk 7 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Harvester.cs]: Harvester is supposed to harvest resources off the ground, transfer them to another trait

File: OpenRA.Mods.Common/Traits/Harvester.cs (hunk 8 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Harvester.cs]: sure

File: OpenRA.Mods.Common/Traits/Harvester.cs (hunk 11 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Harvester.cs:187]: No, just because it's impossible to unload this tick, it doesn't mean that it will be for the next one.

File: OpenRA.Mods.Common/Traits/Harvester.cs (hunk 10 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Harvester.cs:181]: I don't think adjusting makes sense for CnC type resource systems. And Harvester is basically a CnC class / unit.

File: OpenRA.Mods.Common/Traits/StoresPlayerResources.cs (hunk 15 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/StoresPlayerResources.cs:31]: Removed sync

File: OpenRA.Mods.Common/Traits/Render/WithStoresResourcesPipsDecoration.cs (hunk 13 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Render/WithStoresResourcesPipsDecoration.cs]: I think the way it's currently named is the unspoken convention in this codebase

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 20]
```diff
@@ -172,8 +172,31 @@ public interface IVoiced
 		bool HasVoice(Actor self, string voice);
 	}
 
-	[RequireExplicitImplementation]
```
Comment (abcdefg30) [inline OpenRA.Game/Traits/TraitsInterfaces.cs:175]: Why was this removed?

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 21]
```diff
@@ -172,8 +172,31 @@ public interface IVoiced
...
-	public interface IStoreResources { int Capacity { get; } }
+	public interface IStoresResourcesInfo : ITraitInfoInterface
+	{
+		string[] ResourceTypes { get; }
+	}
+
+	public interface IStoresResources
```
Comment (abcdefg30) [inline OpenRA.Game/Traits/TraitsInterfaces.cs:181]: This should also be explicitly implemented.

File: OpenRA.Mods.Common/Traits/StoresPlayerResources.cs (hunk 14 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/StoresPlayerResources.cs:18]: Those should depend on both the trait that adds resources to storage (i.e. Harvester) and the trait that spends it (i.e. Production)

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 22]
```diff
@@ -102,17 +98,35 @@ public class HarvesterInfo : ConditionalTraitInfo, Requires<MobileInfo>
...
+				throw new YamlException($"Harvester.{nameof(Resources)} is empty.");
+
+			foreach (var sr in info.TraitInfos<IStoresResourcesInfo>())
+			{
+				foreach (var type in sr.ResourceTypes)
+				{
+					if (Resources.Contains(type) && resourceTypes.Contains(type))
+						resourceTypes.Remove(type);
+				}
+			}
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Harvester.cs]:
```suggestion
			foreach (var sr in info.TraitInfos<IStoresResourcesInfo>())
				foreach (var type in sr.ResourceTypes)
					if (Resources.Contains(type))
						resourceTypes.Remove(type);
```

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Harvester.cs]: Or something like
```suggestion
			foreach (var type in info.TraitInfos<IStoresResourcesInfo>().SelectMany(sr => sr.ResourceTypes))
				if (Resources.Contains(type))
					resourceTypes.Remove(type);
```
if you prefer that.

File: OpenRA.Game/Traits/TraitsInterfaces.cs (hunk 6 above)
Comment (abcdefg30) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:190]: I don't think the extra complexity is worth it. (At least unless we have some numbers on that this makes a significant impact.)

File: OpenRA.Mods.Common/Traits/StoresResources.cs (hunk 17 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/StoresResources.cs]: good catch

File: OpenRA.Game/Traits/TraitsInterfaces.cs (hunk 20 above)
Comment (PunkPun) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:175]: added back

File: OpenRA.Game/Traits/TraitsInterfaces.cs (hunk 21 above)
Comment (PunkPun) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:181]: Why? It would just force nasty code in other implementations

File: OpenRA.Mods.Common/Traits/Harvester.cs (hunk 22 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Harvester.cs]: made it fully use Linq

File: OpenRA.Mods.Common/Traits/StoresResources.cs (hunk 18 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/StoresResources.cs:53]: There's no sync here?

File: OpenRA.Game/Traits/TraitsInterfaces.cs (hunk 21 above)
Comment (abcdefg30) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:181]: All our interfaces should be explicit. Why would this force nasty code?

Comment (PunkPun) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:181]: because if the interface is explicit, you can't reference members within the class

File: OpenRA.Mods.Common/Traits/StoresResources.cs (hunk 19 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/StoresResources.cs:74]: That would force all resource types to be defined at a world level. Which, granted may be desired.

Regardless this is out of scope of this PR

File: OpenRA.Game/Traits/TraitsInterfaces.cs (hunk 6 above)
Comment (PunkPun) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:190]: It's not a cold path, but also it's not very hot, so such optimisations shouldn't have very noticeable impact. It may make sense to do an optimisation pass after the whole refactor is finished, but atm it would just make the code much harder to read.

File: OpenRA.Mods.Common/Traits/StoresResources.cs (hunk 18 above)
Comment (anvilvapre) [reply OpenRA.Mods.Common/Traits/StoresResources.cs:53]: The line above. Is it really worth while to sync the value of each harvester contineously. And you do this by looping over the contents. I guess there are only a few types of content.

File: OpenRA.Game/Traits/TraitsInterfaces.cs (hunk 6 above)
Comment (anvilvapre) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:190]: 8 haversters times 8 players times any resource update.

File: OpenRA.Mods.Common/Traits/Harvester.cs (hunk 12 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Harvester.cs:269]: I had made sure the division is the last action

File: OpenRA.Mods.Common/Traits/Render/WithStoresResourcesPipsDecoration.cs [hunk 23]
```diff
@@ -10,12 +10,13 @@
...
 
 using System.Collections.Generic;
+using System.Linq;
 using OpenRA.Graphics;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits.Render
 {
-	public class WithHarvesterPipsDecorationInfo : WithDecorationBaseInfo, Requires<HarvesterInfo>
+	public class WithStoresResourcesPipsDecorationInfo : WithDecorationBaseInfo, Requires<IStoresResourcesInfo>
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithStoresResourcesPipsDecoration.cs:19]:
```suggestion
	public class WithResourcesPipsDecorationInfo : WithDecorationBaseInfo, Requires<IStoresResourcesInfo>
```

maybe?

File: OpenRA.Mods.Common/Traits/StoresPlayerResources.cs [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	public class StoresPlayerResources : INotifyOwnerChanged, INotifyCapture, INotifyKilled, INotifyAddedToWorld, INotifyRemovedFromWorld
+	{
+		readonly StoresPlayerResourcesInfo info;
+		PlayerResources player;
+
+		public int Stored => player.ResourceCapacity == 0 ? 0 : (int)((long)info.Capacity * player.Resources / player.ResourceCapacity);
+
+		public StoresPlayerResources(Actor self, StoresPlayerResourcesInfo info)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/StoresPlayerResources.cs:33]:
```suggestion
		public StoresResources(Actor self, StoresResourcesInfo info)
```
is easier to grasp for me.

File: OpenRA.Mods.Common/Traits/StoresResources.cs [hunk 25]
```diff
@@ -9,62 +9,100 @@
...
+
+				return value;
+			}
+		}
+
+		public int ContentsSum { get; private set; } = 0;
+		public IReadOnlyDictionary<string, int> Contents { get; }
+		int IStoresResources.Capacity => info.Capacity;
 
 		public StoresResources(Actor self, StoresResourcesInfo info)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/StoresResources.cs:57]:
```suggestion
		public TransportsResources(Actor self, TransportsResourcesInfo info)
```
then?

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/StoresResources.cs:57]: It wouldn't make sense as this is supposed to be a generic resource storage trait. Not tied specifically to harvesters or harvester-like units

File: OpenRA.Mods.Common/Traits/StoresPlayerResources.cs (hunk 24 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/StoresPlayerResources.cs:33]: Although this matches the existing trait `PlayerResources`.

**Comment by Mailaender** [comment]
In `ts` the Tiberium harvesters will fill up the waste refinery, which they already do on bleed. Nevermind.

**Review by Mailaender — APPROVED** [review]

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ddc98dcc1d631abd9c3963d268df55eafb303ec3...02f1a428c52b7290a12b452539534292d50bcda2)
